.github/workflows/build-test-distribute.yaml

```diff
@@ -120,9 +120,14 @@ jobs:
           retention-days: 7
   distributions:
     needs: ["check", "test", "test_e2e", "test_e2e_env"]
-    if: ${{ always() && !failure() && !cancelled() }}
+    if: ${{ always() }}
     runs-on: ubuntu-latest
     steps:
+      # see https://stackoverflow.com/a/67532120/4907315
+      - name: "Halt due to previous failures"
+        run: |
+          exit 1
+        if: ${{ contains(needs.*.result, 'failure')|| contains(needs.*.result, 'cancelled') }}
```

Comment:
For some reason, `disttributions` still run even if some of `test_e2e`  failed. So we check manually here.
Here is an example of it: https://github.com/jijiechen/kuma/actions/runs/7003299010/attempts/1